feat(aws): add support for Node.js 24.x runtime

casperiv0 opened the pull request:
First contribution to SST, so let me know if there should be more added here 🙏.

casperiv0 committed:
feat(aws): add support for Node.js 24.x runtime

smorimoto commented:
I think there are a few other places that should be changed as well:

```diff
diff --git i/platform/src/components/aws/function.ts w/platform/src/components/aws/function.ts
index d02d29de2..7f283869d 100644
--- i/platform/src/components/aws/function.ts
+++ w/platform/src/components/aws/function.ts
@@ -311,7 +311,7 @@ export interface FunctionArgs {
    * @example
    * ```js
    * {
-   *   runtime: "nodejs22.x"
+   *   runtime: "nodejs24.x"
    * }
    * ```
    */
@@ -319,6 +319,7 @@ export interface FunctionArgs {
     | "nodejs18.x"
     | "nodejs20.x"
     | "nodejs22.x"
+    | "nodejs24.x"
     | "go"
     | "rust"
     | "provided.al2023"
diff --git i/platform/src/components/aws/ssr-site.ts w/platform/src/components/aws/ssr-site.ts
index a1eb78d44..cab112f7a 100644
--- i/platform/src/components/aws/ssr-site.ts
+++ w/platform/src/components/aws/ssr-site.ts
@@ -230,12 +230,12 @@ export interface SsrSiteArgs extends BaseSsrSiteArgs {
      * ```js
      * {
      *   server: {
-     *     runtime: "nodejs22.x"
+     *     runtime: "nodejs24.x"
      *   }
      * }
      * ```
      */
-    runtime?: Input<"nodejs18.x" | "nodejs20.x" | "nodejs22.x">;
+    runtime?: Input<"nodejs18.x" | "nodejs20.x" | "nodejs22.x" | "nodejs24.x">;
     /**
      * The maximum amount of time the server function can run.
      *
```

casperiv0 commented:
Updated, thanks!

dragonautdev commented:
This looks great @casperiv0 and super useful 👍

jamesgibbons92 commented:
this requires being on one of the latest aws provider versions (atleast 7.12.0 I believe) so you can't use this runtime on default sst installation unless you explictly upgrade aws in your provider config. not sure if this is worth mentioning somewhere until the providers are updated in sst

casperiv0 commented:
> this requires being on one of the latest aws provider versions (atleast 7.12.0 I believe) so you can't use this runtime on default sst installation unless you explictly upgrade aws in your provider config. not sure if this is worth mentioning somewhere until the providers are updated in sst

Would it be fine to update the providers directly in SST in this PR? Or are there breaking changes in those updates?

jamesgibbons92 commented:
6 -> 7 is a major update and contains some breaking changes if I recall so
needs to be done carefully

On Mon, 1 Dec 2025, 08:32 Casper Iversen, ***@***.***> wrote:

> *casperiv0* left a comment (sst/sst#6243)
> <https://github.com/sst/sst/pull/6243#issuecomment-3595253397>
>
> this requires being on one of the latest aws provider versions (atleast
> 7.12.0 I believe) so you can't use this runtime on default sst installation
> unless you explictly upgrade aws in your provider config. not sure if this
> is worth mentioning somewhere until the providers are updated in sst
>
> Would it be fine to update the providers directly in SST in this PR? Or
> are there breaking changes in those updates?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sst/sst/pull/6243#issuecomment-3595253397>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQCV6OOSTLASLC2LVVGWW337P4JLAVCNFSM6AAAAACNLYHC7OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJVGI2TGMZZG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

AlperSakarya commented:
@vimtor

vimtor commented:
blocked by #5944

55Cancri commented:
Thanks for working on this @casperiv0! I wanted to add some additional context that might be useful, especially since this is blocked by #5944.

### Relationship with API Gateway Streaming

While researching the Node.js 24 runtime dependency chain, I discovered that a related feature - [API Gateway REST API response streaming](https://aws.amazon.com/about-aws/whats-new/2025/11/api-gateway-response-streaming-rest-apis/) - was announced by AWS on November 19, 2025 (6 days before Node.js 24 runtime on Nov 25).

Both features share a common blocker: they require updating to recent versions of the Pulumi AWS provider.

### Upstream Dependency Timeline

| Date | Event |
|------|-------|
| Nov 19, 2025 | AWS announces API Gateway REST API streaming |
| Nov 21, 2025 | Terraform AWS Provider v6.25.0 released (includes API Gateway streaming support) |
| Nov 25, 2025 | AWS announces Node.js 24 Lambda runtime |
| Dec 9, 2025 | Pulumi releases @pulumi/aws v7.13.0 (based on Terraform v6.25.0) |


- **API Gateway streaming**: Should be available once #5944 merges and SST updates to @pulumi/aws v7.13.0+
- **Node.js 24 runtime**: May require a slightly later Pulumi version (if the `nodejs24.x` enum was added in Terraform v6.26.0+, which came after v6.25.0)

So #5944 is the key blocker here. Once that's merged, SST will be much closer to supporting both Node.js 24 runtime AND API Gateway streaming - both of which are highly requested.

Just sharing this in case the broader context is helpful!

mss-dieliko commented:
Thanks for this nice feature!

Not sure if I may add this here, but what about the fixed node runtime values like here below? Will these runtime versions also be updated?

- https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L711
- https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L848
- https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L930

There is currently no way to override this with configurations/transforms if I'm not mistaken?

vimtor commented:
i'll be reviewing this soon now that the pulumi pr has landed

guylr commented:
> i'll be reviewing this soon now that the pulumi pr has landed



> Thanks for this nice feature!
> 
> Not sure if I may add this here, but what about the fixed node runtime values like here below? Will these runtime versions also be updated?
> 
>     * https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L711
> 
>     * https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L848
> 
>     * https://github.com/anomalyco/sst/blob/5bbaefb153ca7e169699b6397b148d4eef14b7f1/platform/src/components/aws/nextjs.ts#L930
> 
> 
> There is currently no way to override this with configurations/transforms if I'm not mistaken?

Worth looking at this as well if possible. Node.js 20.x in Lambda reaches end-of-life on April 30, 2026

vimtor commented on pkg/runtime/node/build.go:
```diff
@@ -23,6 +23,7 @@ var forceExternal = []string{
 }
 
 var targetMap = map[string]esbuild.Target{
+	"nodejs24.x": esbuild.ES2024,
```
this gives the following error:

```
# github.com/sst/sst/v3/pkg/runtime/node
  ../../pkg/runtime/node/build.go:26:24: undefined: esbuild.ES2024
```

vimtor reviewed (changes requested):
thanks for the contribution @casperiv0 

to make this pr work i had to:

1. update with the `dev` branch
2. upgrade the go `esbuild` package to `v0.27.2` (this gets rid of the error)
3. upgrade the aws pulumi provider version the new major version #6259 

could you take care of 1 and 2? then we will merge this as soon as the aws upgrade pr is merged

casperiv0 committed:
fix: update esbuild

casperiv0 commented:
> thanks for the contribution @casperiv0
> 
> to make this pr work i had to:
> 
> 1. update with the `dev` branch
> 2. upgrade the go `esbuild` package to `v0.27.2` (this gets rid of the error)
> 3. upgrade the aws pulumi provider version the new major version #6259
> 
> could you take care of 1 and 2? then we will merge this as soon as the aws upgrade pr is merged

Thanks, I've updated esbuild and rebased on dev!

devlsh commented:
any progress on this one? would love to get my functions upgraded :)

vimtor commented:
@devlsh the pr #6259 looks very promising, we should be merging it relatively soon

bakeruk commented:
@vimtor do you have an ETA on this one? 🙏🏼

sangharshcs commented:
https://github.com/anomalyco/sst/pull/6387 I have this PR up, if DEV can review it. Recent @pulumi/aws package has support for nodejs24.x runtime.

goto3 commented:
Hello, any update on this?

luizkc commented:
Why is this taking so long? 🤔 Seems like a simple enough change to merge.

vimtor commented:
@luizkc it requires merging https://github.com/anomalyco/sst/pull/6259/ first for it to work

we'll release a major version this week so we can finally merge this as well

rfoel commented:
Let's gooooo!

vimtor commented:
IT HAPPENED

luizkc commented:
so can this be merged now? 👀

vimtor reviewed (approved):
thanks for your contribution @casperiv0 

the esbuild update scared me but i've tried this extensively and checked the release notes and everything should be fine

<details>

<summary>proof of it working</summary>

deployed a bunch of examples:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/f6e0814a-5550-4900-ac04-1b3d7d90ac3c" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/b2319fcc-c042-439b-9b45-984f375113a7" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/44aa6a17-89cd-4de6-8742-b7acea356dc7" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/fdc9d1eb-e71b-4c7a-8623-2cd5143612fa" />
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/c2431e69-64bb-4a5c-9dc0-f4c8ff110085" />

and tested `sst dev`

<img width="2076" height="1086" alt="Image" src="https://github.com/user-attachments/assets/13b31f19-cc43-4e28-86fc-6cafbda728f1" />

</details>

i'll open a separate pr to make node 24 the default for future users

jamiechong commented:
Should this be working with SST `v3`? or only `v4`?

JannikWempe commented:
It is v4 only. It uses the new major version of the AWS Pulumi provider
